File: basics/go-programs/10_crud_app.go
package main

import (
	"encoding/json"
	"fmt"
	"net/http"
	"strconv"
	"sync"
)

/*
	GO PROGRAM 10: SIMPLE CRUD API (In-Memory)
	Concepts:
	- RESTful API structure.
	- JSON encoding/decoding.
	- Using a Map as a database.
	- Mutex for thread-safety.
*/

// Task model
type Task struct {
	ID        int    `json:"id"`
	Title     string `json:"title"`
	Completed bool   `json:"completed"`
}

// In-Memory "Database"
var (
	tasks  = make(map[int]Task)
	nextID = 1
	mu     sync.Mutex // Ensures only one goroutine modifies the map at a time
)

// Add initial data
func init() {
	tasks[nextID] = Task{ID: nextID, Title: "Learn Go", Completed: false}
	nextID++
}

// Handler for /tasks
func tasksHandler(w http.ResponseWriter, r *http.Request) {
	w.Header().Set("Content-Type", "application/json")

	switch r.Method {
	case "GET":
		// READ: List all tasks
		mu.Lock()
		taskList := make([]Task, 0, len(tasks))
		for _, t := range tasks {
			taskList = append(taskList, t)
		}
		mu.Unlock()
		json.NewEncoder(w).Encode(taskList)

	case "POST":
		// CREATE: Add a new task
		var nt Task
		if err := json.NewDecoder(r.Body).Decode(&nt); err != nil {
			http.Error(w, err.Error(), http.StatusBadRequest)
			return
		}
		mu.Lock()
		nt.ID = nextID
		tasks[nextID] = nt
		nextID++
		mu.Unlock()
		w.WriteHeader(http.StatusCreated)
		json.NewEncoder(w).Encode(nt)

	default:
		http.Error(w, "Method not allowed", http.StatusMethodNotAllowed)
	}
}

// Handler for /tasks/{id}
func taskHandler(w http.ResponseWriter, r *http.Request) {
	w.Header().Set("Content-Type", "application/json")
	idStr := r.URL.Path[len("/tasks/"):]
	id, err := strconv.Atoi(idStr)
	if err != nil {
		http.Error(w, "Invalid ID", http.StatusBadRequest)
		return
	}

	mu.Lock()
	defer mu.Unlock()

	task, exists := tasks[id]
	if !exists {
		http.Error(w, "Task not found", http.StatusNotFound)
		return
	}

	switch r.Method {
	case "GET":
		// READ: Single task
		json.NewEncoder(w).Encode(task)

	case "PUT":
		// UPDATE: Update existing task
		var ut Task
		if err := json.NewDecoder(r.Body).Decode(&ut); err != nil {
			http.Error(w, err.Error(), http.StatusBadRequest)
			return
		}
		ut.ID = id
		tasks[id] = ut
		json.NewEncoder(w).Encode(ut)

	case "DELETE":
		// DELETE: Remove task
		delete(tasks, id)
		w.WriteHeader(http.StatusNoContent)

	default:
		http.Error(w, "Method not allowed", http.StatusMethodNotAllowed)
	}
}

func main() {
	http.HandleFunc("/tasks", tasksHandler)
	http.HandleFunc("/tasks/", taskHandler)

	fmt.Println("CRUD Server starting on http://localhost:8082...")
	fmt.Println("Available Endpoints:")
	fmt.Println(" - GET/POST /tasks")
	fmt.Println(" - GET/PUT/DELETE /tasks/{id}")

	http.ListenAndServe(":8082", nil)
}

File: basics/node-programs/10_crud_app.js
const http = require('http');

/*
	NODE.JS PROGRAM 10: SIMPLE CRUD API (In-Memory)
	Concepts:
	- RESTful API structure with Node.js built-ins.
	- JSON parsing and stringifying.
	- Using an Array as an in-memory database.
	- Handling body data in an async/streaming fashion.
*/

// In-Memory Database (equiv for Go's Map)
let tasks = [
	{ id: 1, title: "Learn JavaScript", completed: false }
];
let nextID = 2;

// Utility to parse JSON body from the incoming request stream
async function getBody(req) {
	return new Promise((resolve, reject) => {
		let body = '';
		req.on('data', chunk => body += chunk.toString());
		req.on('end', () => resolve(JSON.parse(body || '{}')));
		req.on('error', err => reject(err));
	});
}

const server = http.createServer(async (req, res) => {
	res.setHeader('Content-Type', 'application/json');

	const urlParts = req.url.split('/');
	const id = parseInt(urlParts[2]); // For /tasks/1

	// ROUTE: /tasks (List/Create)
	if (req.url === '/tasks') {
		if (req.method === 'GET') {
			// READ ALL
			res.end(JSON.stringify(tasks));
		} else if (req.method === 'POST') {
			// CREATE
			try {
				const body = await getBody(req);
				const newTask = { id: nextID++, ...body };
				tasks.push(newTask);
				res.statusCode = 201;
				res.end(JSON.stringify(newTask));
			} catch (err) {
				res.statusCode = 400;
				res.end(JSON.stringify({ error: "Invalid JSON" }));
			}
		}
	} 
	// ROUTE: /tasks/:id (GET/PUT/DELETE)
	else if (urlParts[1] === 'tasks' && id) {
		const taskIndex = tasks.findIndex(t => t.id === id);
		if (taskIndex === -1) {
			res.statusCode = 404;
			res.end(JSON.stringify({ error: "Task not found" }));
			return;
		}

		if (req.method === 'GET') {
			// READ SINGLE
			res.end(JSON.stringify(tasks[taskIndex]));
		} else if (req.method === 'PUT') {
			// UPDATE
			const body = await getBody(req);
			tasks[taskIndex] = { ...tasks[taskIndex], ...body };
			res.end(JSON.stringify(tasks[taskIndex]));
		} else if (req.method === 'DELETE') {
			// DELETE
			tasks.splice(taskIndex, 1);
			res.statusCode = 204;
			res.end();
		}
	} else {
		res.statusCode = 404;
		res.end(JSON.stringify({ error: "Endpoint not found" }));
	}
});

const PORT = 3002;
server.listen(PORT, () => {
	console.log(`CRUD Server starting on http://localhost:${PORT}...`);
	console.log("Try these endpoints:");
	console.log(` - GET/POST http://localhost:${PORT}/tasks`);
	console.log(` - GET/PUT/DELETE http://localhost:${PORT}/tasks/1`);
});
